source/common/ssl/ssl_socket.cc

```diff
@@ -14,8 +14,8 @@ using Envoy::Network::PostIoAction;
 namespace Envoy {
 namespace Ssl {
 
-SslSocket::SslSocket(Context& ctx, InitialState state)
-    : ctx_(dynamic_cast<Ssl::ContextImpl&>(ctx)), ssl_(ctx_.newSsl()) {
+SslSocket::SslSocket(ContextSharedPtr ctx, InitialState state)
+    : ctx_(ctx), ssl_(dynamic_cast<Ssl::ContextImpl&>(*ctx_).newSsl()) {
```

Comment:
Is it possible to make `ctx_` as `std::shared_ptr<Ssl::ContextImpl>` and use `std::dynamic_pointer_cast`? It will eliminates `dynamic_cast`s below.

Reply:
Done